A71-xds-fallback.md

```diff
@@ -0,0 +1,190 @@
+
+### xDS resources caching
+
+The xDS client code caches any responses received from the xDS server. This
+includes:
+- Resources that have been successfully received and can be used.
+- Resources that were recieved but did not pass the validation.
+- Resources that are considered as not existing according
+    to [xDS Protocol Specification][resource-does-not-exist].
+
+Resources that have not been recieved by the time xDS server fallback was
+initiated are not considered cached.
+
+### Initiating xDS Fallback
+
+The fallback process is initiated if the following conditions hold:
+
+* There had been a failure during resource retrieval, as described in [A57]:
+    - Channel reports TRANSIENT_FAILURE.
+    - The ADS stream being closed before the first server response had been
+        received.
+* At least one watcher exists for a resource that is not cached. See above
+    for details on resource caching.
+
+### Changes to the code base
+
+1. Refactor code using XdsClient to support multiple instances. 
+1. Add support for multiple xDS servers in the bootstrap JSON. Current
+    implementation only returns the first entry and ignores the rest.
+1. Add multi-client support to CSDS
+
+#### Refactor code using XdsClient to support multiple instances.
+
+1. Introduce a map of "sub-XdsClient" instances keyed by the data plane
+    target.
+1. For each authority, we need to store an ordered list of channels instead
+    of just one channel.
+1. Whenever we lose contact with a given server and there is at least one
+    resource requested but not cached, if the current server is not the
+    last server in the list, then we create a channel/stream for the next
+    server in the list.
+1. Whenever we get back in contact with a given server, if we have
+    a channel/stream for the next server in the list, close it.
+
+#### Add support for multiple xDS servers in the bootstrap JSON
+
+Currently bootstrap JSON supports multiple xDS servers but semantics are not
+explicitely specified.
+
+1. xDS servers will be attempted in the order they are specified in
+    the bootstrap JSON. Server will only be attempted if the previous entry in
+    the list is not available.
+1. xDS client will report a failure if the last entry in the list is not
+    available.
+
+Currently internal data structures do not allow for more then a single xDS
+server. The implementation needs to be updated to handle multiple servers,
+maintainig their fallback order.
+
+#### Add multi-client support to CSDS
+
+Ses [A40] for details on CSDS.
+
+Add add a field to the ClientConfig message in the CSDS response to indicate
+which channel target the data is associated with.
+
+### Temporary environment variable protection
+
+This option will be behind `GRPC_EXPERIMENTAL_XDS_FALLBACK`. If this variable
+is unset or is false, only one xDS server will be read from the bootstrap
+file. 
+
+## Rationale
+
+Other approaches considered:
```

Comment:
How about a new layer of indirection between the xds client and the authorities?

Each channel would create its own xds client.  That would use shared xds clients responsible for the individual authorities it needs.  It would keep track of what resources were needed from each authority.  If any uncached resources were requested from one authority, it would invalidate all of them and use the fallback resources.  If there were no uncached resources from an authority, and the authority went into fallback mode, then it would keep using the cached resources.

This would allow multiple targets that used the same authority to share the xds connections & much of the fallback behavior (i.e. connecting to a backup and going back to the primary when available).